README.md
jk_php_version_parser
==========

Introduction
------------

This python module implements version parsing as defined by the PHP tool `composer`.

Information about this module can be found here:

* [github.org](https://github.com/jkpubsrc/python-module-jk-php-version-parser)
* [pypi.python.org](https://pypi.python.org/pypi/jk_php_version_parser)

Why this module?
----------------

Python is an excellent language to support management of data, software packages, etc.
Not only PHP `composer` but even some data/software packages provide version information in PHP `composer` notation.
This module has been written to parse this data.

Functionality
--------------------------

This module implements all aspects of the specification provided at:
* https://getcomposer.org/doc/articles/versions.md

This module parses all kinds of valid version notation and produces objects of type `jk_version.BaseVersionConstraint`
(which is the base class for version constraints provided by module `jk_version`). This way this module aids in
validations of versions modeled by instances of `jk_version.Version`.

How to use this module
----------------------

### Import this module

Please include this module into your application by using the following code:

```python
import jk_php_version_parser
```

However, as version and constraint objects are provided by module `jk_version` you will likely need to access classes
provided by `jk_version` as well. Therefore it is recommended to import `jk_version` as well:

```python
import jk_version
```

### Instantiate the parser

Now the first thing you need to do is to instantiate a parser:

```python
parser = jk_php_version_parser.ComposerVersionParser()
```

### Parse a version string

Now the parser is ready for parsing. Here is an example of how to do that:

```python
versionString = ">=0.1.2 <0.9 || 1.0 || 1.0.1 || >1.1.1"
constraint = parser.parse(versionString)
```

Now you received a constraint you can match against a specific version later.

### For debugging purposes: Inspect details about the constraint

However, you might want to first inspect details about this constraint for debugging purposeses. You can easily do this
by serializing the constraint to a JSON data structure. The constraint objects from `jk_version` provide a very suitable
method for this purpose: `toJSON()`. The data returend can then be prettyprinted, e.g. with `jk_json.prettyPrint(..)`:

```python
import jk_json
jk_json.prettyPrint(constraint.toJSON())
```

In this particular example the printed JSON will look like this:

```JSON
[
	"or",
	[
		[
			"and",
			[
				[
					">=",
					"0.1.2"
				],
				[
					"<",
					"0.9"
				]
			]
		],
		[
			"==",
			"1.0"
		],
		[
			"==",
			"1.0.1"
		],
		[
			">",
			"1.1.1"
		]
	]
]
```

### Check versions against the constraint

First of all we need some versions we can check. For demonstration purposes let's create a list of some versions:

```python
versionList = [
	jk_version.Version("0.1.1"),
	jk_version.Version("0.1.2"),
	jk_version.Version("1.0"),
	jk_version.Version("1.0.0"),
	jk_version.Version("1.0.1"),
	jk_version.Version("1.1"),
	jk_version.Version("1.1.1"),
	jk_version.Version("1.2"),
]
```

Now we can check the versions:

```python
for version in versionList:
	print("Version  {:10}: {}".format(
		repr(str(version)),
		"match" if constraint.check(version) else "no match",
	))
```

This will generate the following output:

```
Version  '0.1.1'   : no match
Version  '0.1.2'   : match
Version  '1.0'     : match
Version  '1.0.0'   : match
Version  '1.0.1'   : match
Version  '1.1'     : no match
Version  '1.1.1'   : no match
Version  '1.2'     : match
```

Other Python Modules
-------------------

Compatible module(s):

* [jk_version](https://github.com/jkpubsrc/python-module-jk-version) - Version and constraint objects, version number parsing

Recommended other module(s):

* [jk_json](https://github.com/jkpubsrc/python-module-jk-json) - JSON parser, formatter, pretty-printer
* [jk_prettyprintobj](https://github.com/jkpubsrc/python-module-jk-prettyprintobj) - Mixin for objects for pretty-printing
* [jk_typing](https://github.com/jkpubsrc/python-module-jk-typing) - Argument type checking for function and method calls

Contact Information
-------------------

* Jürgen Knauth: pubsrc@binary-overflow.de

License
-------

This software is provided under the following license:

* Apache Software License 2.0





HISTORY.md
* 2022-01-30
	* Initial commit, first release.



testing/test_parsing.py
#!/usr/bin/python3




import jk_testing
import jk_json
import jk_version

import jk_php_version_parser






PARSER = jk_php_version_parser.ComposerVersionParser()





def _parseAndCompareJSON(ctx:jk_testing.TestContext, textToParse:str, jsonExpected):
	x = PARSER.parse(textToParse)
	assert isinstance(x, jk_version.BaseVersionConstraint)
	jsonReceived = x.toJSON()

	linesReceived = jk_json.dumps(jsonReceived, indent="\t", sort_keys=True).split("\n")
	linesExpected = jk_json.dumps(jsonExpected, indent="\t", sort_keys=True).split("\n")

	if jsonReceived == jsonExpected:
		ctx.log.debug("--input---" + "-" * 90)
		ctx.log.debug(textToParse)
		ctx.log.debug("--result--" + "-" * 90)
		for line in linesReceived:
			ctx.log.debug(line)
		ctx.log.debug("-" * 100)
		return

	ctx.log.warn("--input---" + "-" * 90)
	ctx.log.warn(textToParse)
	ctx.log.warn("--result--" + "-" * 90)
	for line in linesReceived:
		ctx.log.warn(line)
	ctx.log.warn("--expected" + "-" * 90)
	for line in linesExpected:
		ctx.log.warn(line)
	ctx.log.warn("-" * 100)

	raise jk_testing.AssertionException("received != expected!")
#






#
# Successes
#

@jk_testing.TestCase()
def test_parsing_elementar_1(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">= 1.2.3",
		[	">=",	"1.2.3"	]
	)
#

@jk_testing.TestCase()
def test_parsing_elementar_caret(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "^1.2.3",
		[
			"and",
			[
				[	">=",	"1.2.3"	],
				[	"<",	"2.0.0"	],
			],
		]
	)
#

@jk_testing.TestCase()
def test_parsing_elementar_tilde(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "~1.2.3",
		[
			"and",
			[
				[	">=",	"1.2.3"	],
				[	"<",	"1.3.0"	],
			],
		]
	)
#

@jk_testing.TestCase()
def test_parsing_multi_1(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.2.3 <1.4",
		[
			"and",
			[
				[	">=",	"1.2.3"	],
				[	"<",	"1.4"	],
			],
		]
	)
#

@jk_testing.TestCase()
def test_parsing_multi_or(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "1.2.3 || 1.2.5",
		[
			"or",
			[
				[	"==", "1.2.3",	],
				[	"==", "1.2.5",	],
			],
		]
	)
#

#
# ----
# Test Cases https://getcomposer.org/doc/articles/versions.md
# ----
#

@jk_testing.TestCase()
def test_parsing_composer_1(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "1.3.2",
		[	"==", "1.3.2",	],
	)
#

@jk_testing.TestCase()
def test_parsing_composer_2(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.3.2",
		[	">=", "1.3.2",	],
	)
#

@jk_testing.TestCase()
def test_parsing_composer_3(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "<1.3.2",
		[	"<", "1.3.2",	],
	)
#

@jk_testing.TestCase()
def test_parsing_composer_4a(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.3.0 <1.4.0",
		[
			"and",
			[
				[	">=", "1.3.0",	],
				[	"<", "1.4.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_4b(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "1.3.*",
		[
			"and",
			[
				[	">=", "1.3.0",	],
				[	"<", "1.4.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_5a(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.3.2 <1.4.0",
		[
			"and",
			[
				[	">=", "1.3.2",	],
				[	"<", "1.4.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_5b(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "~1.3.2",
		[
			"and",
			[
				[	">=", "1.3.2",	],
				[	"<", "1.4.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_6a(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.3.0 <2.0.0",
		[
			"and",
			[
				[	">=", "1.3.0",	],
				[	"<", "2.0.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_6b(ctx:jk_testing.TestContext):
	x = PARSER.parse("~1.3")
	_parseAndCompareJSON(ctx, "~1.3",
		[
			"and",
			[
				[	">=", "1.3",	],
				[	"<", "2.0",		],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_7a(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=1.3.2 <2.0.0",
		[
			"and",
			[
				[	">=", "1.3.2",	],
				[	"<", "2.0.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_7b(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "^1.3.2",
		[
			"and",
			[
				[	">=", "1.3.2",	],
				[	"<", "2.0.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_8a(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, ">=0.3.2 <0.4.0",
		[
			"and",
			[
				[	">=", "0.3.2",	],
				[	"<", "0.4.0",	],
			]
		]
	)
#

@jk_testing.TestCase()
def test_parsing_composer_8b(ctx:jk_testing.TestContext):
	_parseAndCompareJSON(ctx, "^0.3.2",
		[
			"and",
			[
				[	">=", "0.3.2",	],
				[	"<", "0.4.0",	],
			]
		]
	)
#

#
# Errors
#


# TODO



################################################################################################################################



testDriver = jk_testing.TestDriver()

results = testDriver.runTests([
	(test_parsing_elementar_1, True),
	(test_parsing_elementar_caret, True),
	(test_parsing_elementar_tilde, True),

	(test_parsing_multi_1, True),
	(test_parsing_multi_or, True),

	(test_parsing_composer_1, True),
	(test_parsing_composer_2, True),
	(test_parsing_composer_3, True),
	(test_parsing_composer_4a, True),
	(test_parsing_composer_4b, True),
	(test_parsing_composer_5a, True),
	(test_parsing_composer_5b, True),
	(test_parsing_composer_6a, True),
	(test_parsing_composer_6b, True),
	(test_parsing_composer_7a, True),
	(test_parsing_composer_7b, True),
	(test_parsing_composer_8a, True),
	(test_parsing_composer_8b, True),
])

reporter = jk_testing.TestReporterHTML()
reporter.report(results, webbrowserType="chromium")













testing/test_tokenizing.py
#!/usr/bin/python3




import jk_testing
import jk_version

import jk_php_version_parser





tokenizer = jk_php_version_parser.ComposerVersionTokenizer()





#
# Successes
#

@jk_testing.TestCase()
def test_tokenizing_version_1(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 1)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_version_2(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("1.2.3 2.3.4")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
			jk_php_version_parser._ComposerTokenPattern("v", "2.3.4"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_ge(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize(">= 1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", ">="),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_gt(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("> 1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", ">"),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_le(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("<= 1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", "<="),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_lt(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("< 1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", "<"),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_range(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("1.2.3 - 2.3.4")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 3)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
			jk_php_version_parser._ComposerTokenPattern("op", "-"),
			jk_php_version_parser._ComposerTokenPattern("v", "2.3.4"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_version_wildcard(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("1.2.*")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 1)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.*"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_version_gt_wildcard(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize(">=1.2.*")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", ">="),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.*"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_version_caret(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("^1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", "^"),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#

@jk_testing.TestCase()
def test_tokenizing_version_tilde(ctx:jk_testing.TestContext):
	tokenStream = tokenizer.tokenize("~1.2.3")
	tokenStream.dump(printFunc=ctx.log.debug)

	jk_testing.Assert.isEqual(len(tokenStream), 2)

	jk_testing.Assert.isNotNone(
		tokenStream.tryMatchSequence(
			jk_php_version_parser._ComposerTokenPattern("op", "~"),
			jk_php_version_parser._ComposerTokenPattern("v", "1.2.3"),
		)
	)
#




#
# Errors
#


# TODO



################################################################################################################################



testDriver = jk_testing.TestDriver()

results = testDriver.runTests([
	(test_tokenizing_version_1, True),
	(test_tokenizing_version_2, True),
	(test_tokenizing_ge, True),
	(test_tokenizing_gt, True),
	(test_tokenizing_le, True),
	(test_tokenizing_lt, True),
	(test_tokenizing_range, True),
	(test_tokenizing_version_wildcard, True),
	(test_tokenizing_version_gt_wildcard, True),
	(test_tokenizing_version_caret, True),
	(test_tokenizing_version_tilde, True),
	#(test_OR, True),
	#(test_AND, True),
])

reporter = jk_testing.TestReporterHTML()
reporter.report(results, webbrowserType="chromium")













examples/example_1.py
#!/usr/bin/python3


import jk_version
import jk_php_version_parser
import jk_json



# ----------------------------------------------------------------
# Instantiate the parser

parser = jk_php_version_parser.ComposerVersionParser()

# ----------------------------------------------------------------
# Parse a version string

versionString = ">=0.1.2 <0.9 || 1.0 || 1.0.1 || >1.1.1"
constraint = parser.parse(versionString)

# ----------------------------------------------------------------
# For debugging purposes: Inspect details about the constraint

jk_json.prettyPrint(constraint.toJSON())

print()
print("-" * 120)
print()

# ----------------------------------------------------------------
# Check versions against the constraint

versionList = [
	jk_version.Version("0.1.1"),
	jk_version.Version("0.1.2"),
	jk_version.Version("1.0"),
	jk_version.Version("1.0.0"),
	jk_version.Version("1.0.1"),
	jk_version.Version("1.1"),
	jk_version.Version("1.1.1"),
	jk_version.Version("1.2"),
]

for version in versionList:
	print("Version  {:10}: {}".format(
		repr(str(version)),
		"match" if constraint.check(version) else "no match",
	))



